### PR TITLE
Feature/light shield

### PR DESCRIPTION
**Pull request by Dartemiss**
What's new?
- Light shield implemented! It stops damage from enemies and makes the lightball rebound. For activate the light shield use "K" on a keyboard or (Triangle/Y) on a controller.
- Knockback is now a parameter by default of the TakeDamage function, and each attack has a different knockback power value.
- Stun has been implemented for the player, when lightshield breaks, the player gets stunned by a certain time.
- Visuals for shield are placeholders (using same as lightorbs) and a stunning effect from the internet.

PD: For trying to balance the lightball rebound I implemented some optional features but test it, please. On the LightBall script, there is a Checkbox called "Use parry for rebound" that will allow rebound if u parry on the correct moment.
Also on the PlayerController there is a checkbox called "Slow with shield" that will slow the player movement speed when using the shield.

### CODE REVIEW

File: Models/Male Character/ShieldEffect.prefab [hunk 1]
```diff
@@ -0,0 +1,349 @@
+{
+    "GameObjects": [
```
Comment (yalania) [inline Models/Male Character/ShieldEffect.prefab:2]: The prefab!

File: Scripts/src/Script/LightShield.cpp [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+LightShield* LightShieldDLL()
+{
+	LightShield* instance = new LightShield();
+	return instance;
+}
+
+
+LightShield::LightShield()
+{
+
```
Comment (yalania) [inline Scripts/src/Script/LightShield.cpp:28]: LightShield() = default on header?

File: Scripts/src/Script/LightBall.cpp [hunk 3]
```diff
@@ -390,6 +404,16 @@ bool LightBall::AlreadyHitted(EnemyController* enemy)
 	return std::find(enemies_hitted.begin(), enemies_hitted.end(), enemy) != enemies_hitted.end();
 }
 
+void LightBall::SwapPlayersTarget()
+{
+	if (world_manager->multiplayer)
+	{
+		std::swap(current_player, next_target);
```
Comment (yalania) [inline Scripts/src/Script/LightBall.cpp:411]: std::swap :O niceee

File: Scripts/src/Script/LightShield.cpp (hunk 2 above)
Comment (Dartemiss) [reply Scripts/src/Script/LightShield.cpp:28]: yes! done :D

File: Scripts/src/Script/LightBall.cpp [hunk 4]
```diff
@@ -103,7 +104,7 @@ void LightBall::Update()
 
 	if(input_manager->GetGameInput("LightBall", static_cast<PlayerID>(current_player_controller->player - 1)))
 	{
-		if(!moving && CanBeActivated() && current_cooldown <= 0 && current_state == LightBallState::UNNACTIVE)
+		if(!moving && CanBeActivated() && current_cooldown <= 0 && current_state == LightBallState::UNNACTIVE && !current_player_controller->light_shield->active)
```
Comment (mariofv) [inline Scripts/src/Script/LightBall.cpp:107]: This condition is starting to be complex and big, could it be possible to encapsulate it in a method?

File: Scripts/src/Script/LightBall.cpp [hunk 5]
```diff
@@ -228,24 +229,36 @@ void LightBall::Update()
...
 				current_cooldown = cooldown;
+				speed = default_speed;
+				maximum_time_to_return = default_maximum_time_to_return;
 
 				//TODO Delete this when OnCollisionEnter implemented
 				enemies_hitted.clear();
 				//Debug
 				transition = false;
 			}
+			else if(((use_parry_for_rebound) ? current_player_controller->light_shield->parry_enable : current_player_controller->light_shield->active)  && ball_distance_magnitude < shield_radius_range)
```
Comment (mariofv) [inline Scripts/src/Script/LightBall.cpp:247]: This is quite difficult to read, separate it in different lines. Also having a ternary operator that returns a boolean in the condition is weird, probably the logic could be rewritten in a different way

File: Scripts/src/Script/PlayerController.cpp [hunk 6]
```diff
@@ -126,12 +130,41 @@ void PlayerController::Update()
 		case PlayerState::ALIVE:
 			if(!is_reviving)
 			{
```
Comment (mariofv) [inline Scripts/src/Script/PlayerController.cpp:132]: This  method is starting to get big, we need to encapsulate behaviours in methods for readability and faster debugging

File: Scripts/src/Script/LightBall.cpp (hunk 5 above)
Comment (Dartemiss) [reply Scripts/src/Script/LightBall.cpp:247]: Is was just for debug and letting you test it with parry and without it, once we decide which one we will use ill change the ternary operator by what we choose.

File: Scripts/src/Script/LightBall.cpp (hunk 4 above)
Comment (Dartemiss) [reply Scripts/src/Script/LightBall.cpp:107]: Done!

File: Scripts/src/Script/LightShield.cpp [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+// void LightShield::Load(const Config& config)
+
+// {
+
+// 	exampleUUID = config.GetUInt("ExampleNameforSave", 0);
+
+// 	Script::Load(config);
+
+// }
```
Comment (JorxSS) [inline Scripts/src/Script/LightShield.cpp:227]: Delete if not used

File: Scripts/src/Script/LightShield.h [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+	~LightShield() = default;
+
+	void Awake() override;
+	void Start() override;
+	void Update() override;
+
+	void OnInspector(ImGuiContext*) override;
+	void InitPublicGameObjects();
+	//void Save(Config& config) const override;
+	//void Load(const Config& config) override;
```
Comment (JorxSS) [inline Scripts/src/Script/LightShield.h:20]: same

File: Scripts/src/Script/LightShield.cpp (hunk 7 above)
Comment (Dartemiss) [reply Scripts/src/Script/LightShield.cpp]: Done!

**Pull request merged**